File: utils/scripts/helpers/type-mapping.ts
/**
 * Type mapping utilities for converting JSON Schema types to TypeScript types
 */

import type { SchemaProperty } from '../types';

/**
 * Maps JSON Schema primitive types to TypeScript types
 * @param type - JSON Schema type (string, integer, number, boolean, null)
 * @returns TypeScript type as string
 */
export function mapJsonTypeToTs(type: string): string {
  switch (type) {
    case 'string': return 'string';
    case 'integer': return 'number';
    case 'number': return 'number';
    case 'boolean': return 'boolean';
    case 'null': return 'null';
    default: return 'any';
  }
}

/**
 * Converts a JSON Schema property definition to a TypeScript type string
 * Handles:
 * - Date/DateTime/Time formats (converts to Date type)
 * - Nullable types (union with null)
 * - Arrays
 * - Primitive types
 *
 * @param prop - JSON Schema property definition
 * @returns TypeScript type as string
 */
export function getTsType(prop: SchemaProperty, forViewProps: boolean = false): string {
  // Check for date/datetime/time format
  const format = (prop as any).format;
  const isDateType = format === 'date' || format === 'date-time' || format === 'time';

  if (Array.isArray(prop.type)) {
    // Union type for nullable
    if (isDateType) {
      // For nullable date types
      return prop.type.includes('null') || forViewProps ? 'Date | null' : 'Date';
    }
    return prop.type.map(t => t === 'null' ? 'null' : mapJsonTypeToTs(t)).join(' | ');
  }

  if (prop.type === 'array') {
    return 'any[]'; // Simplified
  }

  if (isDateType && prop.type === 'string') {
    return forViewProps ? 'Date | null' : 'Date';
  }

  return mapJsonTypeToTs(prop.type);
}

File: code_generator/helpers/type_mapping.py
"""JSON Schema → TypeScript type mapping — port of helpers/type-mapping.ts"""


def _map_primitive(t: str) -> str:
    return {'string': 'string', 'integer': 'number', 'number': 'number',
            'boolean': 'boolean', 'null': 'null'}.get(t, 'any')


def get_ts_type(prop: dict, for_view_props: bool = False) -> str:
    fmt = prop.get('format')
    is_date = fmt in ('date', 'date-time', 'time')
    prop_type = prop.get('type')

    # A Prisma nativeEnum-backed field (schema_deriver._prisma_native_enum_type
    # marker) has a concrete string-literal-union type on the Prisma client
    # side (e.g. 'pending' | 'rejected'), narrower than plain `string`. Emit
    # that literal union so values assigned into a Prisma `data: {...}` write
    # type-check without a cast. The exposed JSON `type` is unaffected
    # (still "string" -- Class B: API/JSON shape unchanged).
    if prop.get('_prisma_native_enum_type') and prop.get('enum') and prop_type == 'string':
        union = ' | '.join(f"'{v}'" for v in prop['enum'])
        return f"{union} | null" if for_view_props else union

    if isinstance(prop_type, list):
        if is_date:
            return 'Date | null' if ('null' in prop_type or for_view_props) else 'Date'
        return ' | '.join('null' if t == 'null' else _map_primitive(t) for t in prop_type)

    if prop_type and ('number' in prop_type or 'integer' in prop_type):
        return 'number | null' if ('null' in prop_type or for_view_props) else 'number'

    if prop_type == 'array':
        return 'any[]'

    if is_date and prop_type == 'string':
        return 'Date | null' if for_view_props else 'Date'

    return _map_primitive(prop_type or '')
